File: milk2.cpp
/*
ID: totheso1
LANG: C++14
TASK: milk2
*/

#include <algorithm>
#include <fstream>
#include <iterator>
#include <utility>
#include <vector>

using namespace std;

static ifstream fin("milk2.in");
static ofstream fout("milk2.out");

template <typename T>
T fin_get() {
  return *istream_iterator<T>(fin);
}

template <typename C>
C fin_get_collection(int size) {
  C ret;
  copy_n(istream_iterator<typename C::value_type>(fin), size,
         back_inserter(ret));
  return ret;
}

using milking_event = pair<int, bool>;

int main() {
  vector<milking_event> events;
  for (int number_of_milkings = fin_get<int>(); number_of_milkings;
       --number_of_milkings) {
    events.emplace_back(fin_get<int>(), false);
    events.emplace_back(fin_get<int>(), true);
  }
  sort(begin(events), end(events));
  int longest_milking = 0;
  int longest_idleing = 0;
  int start = 0;
  int end = 0;
  int milking = 0;
  for (const milking_event &event : events) {
    const int time = event.first;
    const bool farmer_leaving = event.second;
    if (milking == 0 && !farmer_leaving) {
      start = time;
      if (end != 0) {
        longest_idleing = max(longest_idleing, start - end);
      }
    }
    if (milking == 1 && farmer_leaving) {
      end = time;
      longest_milking = max(longest_milking, end - start);
    }
    milking -= farmer_leaving * 2 - 1;
  }
  fout << longest_milking << ' ' << longest_idleing << endl;
}

File: milk2.py
"""
ID: totheso1
LANG: PYTHON3
TASK: milk2
"""

fin = open("milk2.in", "r")
fout = open("milk2.out", "w")


def fprint(*args, **kwargs) -> None:
    print(*args, file=fout, **kwargs)


# Code start
# Simple event counting with bucket sort, O(t), start
# TIME_MAX = 1000000
# new_beginner_count_by_time = [0] * TIME_MAX
# number_of_milkings = int(fin.readline())
# for line in fin:
#     start, end = map(int, line.split())
#     new_beginner_count_by_time[start] += 1
#     new_beginner_count_by_time[end] -= 1

# longest_milking = longest_idling = 0
# start = end = 0
# milking = 0
# for i, count in enumerate(new_beginner_count_by_time):
#     if milking == 0 and count:
#         start = i
#         if end:
#             longest_idling = max(longest_idling, start - end)
#     if milking and milking == -count:
#         end = i
#         longest_milking = max(longest_milking, end - start)
#     milking += count

# fprint(longest_milking, longest_idling)
# Simple event counting with bucket sort, O(N + t), end

# Same event counting with discretization and system sort, O(NlogN), start
number_of_milkings = int(fin.readline())
events = []
for line in fin:
    start, end = map(int, line.split())
    events.append((start, False))
    events.append((end, True))
events.sort()

longest_milking = longest_idling = 0
start = end = 0
milking = 0
for time, farmer_leaving in events:
    if milking == 0 and not farmer_leaving:
        start = time
        if end:
            longest_idling = max(longest_idling, start - end)
    if milking == 1 and farmer_leaving:
        end = time
        longest_milking = max(longest_milking, end - start)
    milking -= farmer_leaving * 2 - 1

fprint(longest_milking, longest_idling)
# Same event counting with discretization and system sort, O(NlogN), end
# Code end
